File: core/py/src/braintrust_core/span_identifier.py
# Serialization format for capturing all relevant information about a span
# necessary for distributed logging / tracing. Meant to be passed around as an
# opaque string.

import base64
import dataclasses
from enum import Enum, auto
from typing import Optional
from uuid import UUID


def _try_make_uuid(s):
    try:
        ret = UUID(s).bytes
        assert len(ret) == 16
        return ret, True
    except Exception:
        return s.encode("utf-8"), False


ENCODING_VERSION_NUMBER = 1

INVALID_ENCODING_ERRMSG = "SpanComponents string is not properly encoded. This may be due to a version mismatch between the SDK library used to export the span and the library used to decode it. Please make sure you are using the same SDK version across the board"


class SpanObjectType(Enum):
    EXPERIMENT = auto()
    PROJECT_LOGS = auto()

    def __str__(self):
        return {SpanObjectType.EXPERIMENT: "experiment", SpanObjectType.PROJECT_LOGS: "project_logs"}[self]


@dataclasses.dataclass
class SpanRowIds:
    row_id: str
    span_id: str
    root_span_id: str

    def __post_init__(self):
        assert isinstance(self.row_id, str)
        assert isinstance(self.span_id, str)
        assert isinstance(self.root_span_id, str)
        assert self.row_id
        assert self.span_id
        assert self.root_span_id


@dataclasses.dataclass
class SpanComponents:
    object_type: SpanObjectType
    object_id: str
    row_ids: Optional[SpanRowIds] = None

    def __post_init__(self):
        assert isinstance(self.object_type, SpanObjectType)
        assert isinstance(self.object_id, str)
        if self.row_ids is not None:
            assert isinstance(self.row_ids, SpanRowIds)

    def to_str(self) -> str:
        # Our binary object format is as follows:
        #   - Byte 0 encodes the version number of the encoded string. This is
        #   used to check for incompatibilities with previous iterations.
        #   - Byte 1 encodes the SpanObjectType.
        #   - Byte 2 (0 or 1) describes whether or not the (row_id,
        #   span_id, root_span_id) triple is present.
        #   - Byte 3 (0 or 1) describes whether or not the row_id component
        #   is a UUID. If not, it is assumed to be a utf-8 encoded string.
        #   - Bytes 4-19 encode the object_id as a UUID
        #   - If the row triple is present, bytes 20-51 encode the span_id +
        #   root_span_id as a UUID.
        #   - If the row triple is present, the remaining bytes encode the
        #   row_id either as UUID or as UTF-8.

        if self.row_ids:
            row_id_bytes, row_id_is_uuid = _try_make_uuid(self.row_ids.row_id)
        else:
            row_id_bytes, row_id_is_uuid = None, False

        raw_bytes = bytes(
            [
                ENCODING_VERSION_NUMBER,
                self.object_type.value,
                1 if self.row_ids else 0,
                1 if row_id_is_uuid else 0,
            ]
        )

        object_id_bytes, object_id_is_uuid = _try_make_uuid(self.object_id)
        if not object_id_is_uuid:
            raise Exception("object_id component must be a valid UUID")
        raw_bytes += object_id_bytes

        if self.row_ids:
            span_id_bytes, span_id_is_uuid = _try_make_uuid(self.row_ids.span_id)
            if not span_id_is_uuid:
                raise Exception("span_id component must be a valid UUID")
            root_span_id_bytes, root_span_id_is_uuid = _try_make_uuid(self.row_ids.root_span_id)
            if not root_span_id_is_uuid:
                raise Exception("root_span_id component must be a valid UUID")
            raw_bytes += span_id_bytes
            raw_bytes += root_span_id_bytes
            raw_bytes += row_id_bytes

        return base64.b64encode(raw_bytes).decode()

    @staticmethod
    def from_str(s: str) -> "SpanComponents":
        try:
            raw_bytes = base64.b64decode(s.encode())
            assert raw_bytes[0] == ENCODING_VERSION_NUMBER
            object_type = SpanObjectType(raw_bytes[1])
            assert raw_bytes[2] in [0, 1]
            assert raw_bytes[3] in [0, 1]
            has_row_id = raw_bytes[2] == 1
            row_id_is_uuid = raw_bytes[3] == 1

            object_id = str(UUID(bytes=raw_bytes[4:20]))
            if has_row_id:
                span_id = str(UUID(bytes=raw_bytes[20:36]))
                root_span_id = str(UUID(bytes=raw_bytes[36:52]))
                if row_id_is_uuid:
                    row_id = str(UUID(bytes=raw_bytes[52:]))
                else:
                    row_id = raw_bytes[52:].decode("utf-8")
                row_ids = SpanRowIds(row_id=row_id, span_id=span_id, root_span_id=root_span_id)
            else:
                row_ids = None

            return SpanComponents(object_type=object_type, object_id=object_id, row_ids=row_ids)
        except Exception:
            raise Exception(INVALID_ENCODING_ERRMSG)

    def object_id_fields(self):
        if self.object_type == SpanObjectType.EXPERIMENT:
            return dict(experiment_id=self.object_id)
        elif self.object_type == SpanObjectType.PROJECT_LOGS:
            return dict(project_id=self.object_id, log_id="g")

File: core/js/src/span_identifier.ts
// Mirror of core/py/src/braintrust_core/span_parent_identifier.py.

import * as uuid from "uuid";
import { ParentExperimentIds, ParentProjectLogIds } from "./object";
import { z } from "zod";

function tryMakeUuid(s: string): { bytes: Buffer; isUUID: boolean } {
  try {
    const ret = uuid.parse(s);
    if (ret.length !== 16) {
      throw new Error();
    }
    return { bytes: Buffer.from(ret), isUUID: true };
  } catch (e) {
    return { bytes: Buffer.from(s, "utf-8"), isUUID: false };
  }
}

const ENCODING_VERSION_NUMBER = 2;

const INVALID_ENCODING_ERRMSG =
  "SpanComponents string is not properly encoded. This may be due to a version mismatch between the SDK library used to export the span and the library used to decode it. Please make sure you are using the same SDK version across the board";
// If you change this, make sure to change the method used to read/write integer
// bytes to a buffer, from writeInt32BE.
const INTEGER_ENCODING_NUM_BYTES = 4;

export enum SpanObjectType {
  EXPERIMENT = 0,
  PROJECT_LOGS = 1,
}

const SpanObjectTypeEnumSchema = z.nativeEnum(SpanObjectType);

export class SpanRowIds {
  public rowId: string;
  public spanId: string;
  public rootSpanId: string;

  constructor(args: { rowId: string; spanId: string; rootSpanId: string }) {
    this.rowId = args.rowId;
    this.spanId = args.spanId;
    this.rootSpanId = args.rootSpanId;

    if (!this.rowId) {
      throw new Error("rowId must be nonempty string");
    }
    if (!this.spanId) {
      throw new Error("spanId must be nonempty string");
    }
    if (!this.rootSpanId) {
      throw new Error("rootSpanId must be nonempty string");
    }
  }

  public toObject() {
    return {
      rowId: this.rowId,
      spanId: this.spanId,
      rootSpanId: this.rootSpanId,
    };
  }
}

export class SpanComponents {
  public objectType: SpanObjectType;
  public objectId: string | undefined;
  public computeObjectMetadataArgs: Record<string, any> | undefined;
  public rowIds: SpanRowIds | undefined;

  constructor(args: {
    objectType: SpanObjectType;
    objectId?: string;
    computeObjectMetadataArgs?: Record<string, any>;
    rowIds?: SpanRowIds;
  }) {
    this.objectType = args.objectType;
    this.objectId = args.objectId;
    this.computeObjectMetadataArgs = args.computeObjectMetadataArgs;
    this.rowIds = args.rowIds;

    if (!(this.objectId || this.computeObjectMetadataArgs)) {
      throw new Error(
        "Must provide either objectId or computeObjectMetadataArgs",
      );
    }
  }

  public toStr(): string {
    const allBuffers: Array<Buffer> = [];

    const { bytes: rowIdBytes, isUUID: rowIdIsUUID } = this.rowIds
      ? tryMakeUuid(this.rowIds.rowId)
      : { bytes: Buffer.from(""), isUUID: false };

    allBuffers.push(
      Buffer.from([
        ENCODING_VERSION_NUMBER,
        this.objectType,
        this.objectId ? 1 : 0,
        this.computeObjectMetadataArgs ? 1 : 0,
        this.rowIds ? 1 : 0,
        rowIdIsUUID ? 1 : 0,
      ]),
    );

    if (this.objectId) {
      const { bytes: objectIdBytes, isUUID: objectIdIsUUID } = tryMakeUuid(
        this.objectId,
      );
      if (!objectIdIsUUID) {
        throw new Error("object_id component must be a valid UUID");
      }
      allBuffers.push(objectIdBytes);
    }

    if (this.computeObjectMetadataArgs) {
      const computeObjectMetadataBytes = Buffer.from(
        JSON.stringify(this.computeObjectMetadataArgs),
        "utf-8",
      );
      const serializedLenBytes = Buffer.alloc(INTEGER_ENCODING_NUM_BYTES);
      serializedLenBytes.writeInt32BE(computeObjectMetadataBytes.length);
      allBuffers.push(serializedLenBytes, computeObjectMetadataBytes);
    }

    if (this.rowIds) {
      const { bytes: spanIdBytes, isUUID: spanIdIsUUID } = tryMakeUuid(
        this.rowIds.spanId,
      );
      if (!spanIdIsUUID) {
        throw new Error("span_id component must be a valid UUID");
      }
      const { bytes: rootSpanIdBytes, isUUID: rootSpanIdIsUUID } = tryMakeUuid(
        this.rowIds.rootSpanId,
      );
      if (!rootSpanIdIsUUID) {
        throw new Error("root_span_id component must be a valid UUID");
      }
      allBuffers.push(spanIdBytes, rootSpanIdBytes, rowIdBytes);
    }

    return Buffer.concat(allBuffers).toString("base64");
  }

  public static fromStr(s: string): SpanComponents {
    try {
      const rawBytes = Buffer.from(s, "base64");
      if (rawBytes[0] !== ENCODING_VERSION_NUMBER) {
        throw new Error();
      }
      const objectType = SpanObjectTypeEnumSchema.parse(rawBytes[1]);
      for (let i = 2; i < 6; ++i) {
        if (![0, 1].includes(rawBytes[i])) {
          throw new Error();
        }
      }
      const hasObjectId = rawBytes[2] == 1;
      const hasComputeObjectMetadataArgs = rawBytes[3] == 1;
      const hasRowId = rawBytes[4] == 1;
      const rowIdIsUUID = rawBytes[5] == 1;

      let byteCursor = 6;
      let objectId: string | undefined = undefined;
      if (hasObjectId) {
        const nextByteCursor = byteCursor + 16;
        objectId = uuid.stringify(
          rawBytes.subarray(byteCursor, nextByteCursor),
        );
        byteCursor = nextByteCursor;
      }

      let computeObjectMetadataArgs: Record<string, any> | undefined;
      if (hasComputeObjectMetadataArgs) {
        let nextByteCursor = byteCursor + INTEGER_ENCODING_NUM_BYTES;
        const serializedLenBytes = rawBytes.readInt32BE(byteCursor);
        byteCursor = nextByteCursor;
        nextByteCursor = byteCursor + serializedLenBytes;
        computeObjectMetadataArgs = JSON.parse(
          rawBytes.subarray(byteCursor, nextByteCursor).toString("utf-8"),
        );
        byteCursor = nextByteCursor;
      }

      const rowIds = (() => {
        if (!hasRowId) {
          return undefined;
        }
        let nextByteCursor = byteCursor + 16;
        const spanId = uuid.stringify(
          rawBytes.subarray(byteCursor, nextByteCursor),
        );
        byteCursor = nextByteCursor;
        nextByteCursor = byteCursor + 16;
        const rootSpanId = uuid.stringify(
          rawBytes.subarray(byteCursor, nextByteCursor),
        );
        byteCursor = nextByteCursor;
        const rowId = rowIdIsUUID
          ? uuid.stringify(rawBytes.subarray(byteCursor))
          : rawBytes.subarray(byteCursor).toString("utf-8");
        return new SpanRowIds({ rowId, spanId, rootSpanId });
      })();

      return new SpanComponents({
        objectType,
        objectId,
        computeObjectMetadataArgs,
        rowIds,
      });
    } catch (e) {
      throw new Error(INVALID_ENCODING_ERRMSG);
    }
  }

  public objectIdFields(): ParentExperimentIds | ParentProjectLogIds {
    if (!this.objectId) {
      throw new Error(
        "Impossible: cannot invoke `object_id_fields` unless SpanComponents is initialized with an `object_id`",
      );
    }
    switch (this.objectType) {
      case SpanObjectType.EXPERIMENT:
        return { experiment_id: this.objectId };
      case SpanObjectType.PROJECT_LOGS:
        return { project_id: this.objectId, log_id: "g" };
      default:
        throw new Error("Impossible");
    }
  }

  public toObject() {
    return {
      objectType: this.objectType,
      objectId: this.objectId,
      computeObjectMetadataArgs: this.computeObjectMetadataArgs,
      rowIds: this.rowIds?.toObject(),
    };
  }
}
